types/src/vm/modules.rs

```diff
@@ -0,0 +1,172 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    delayed_fields::PanicError,
+    state_store::state_value::{StateValue, StateValueMetadata},
+    write_set::TransactionWrite,
+};
+use bytes::Bytes;
+use move_binary_format::{errors::VMResult, CompiledModule};
+use move_core_types::metadata::Metadata;
+use move_vm_runtime::{Module, RuntimeEnvironment};
+use std::{fmt::Debug, sync::Arc};
+
+/// An interface that any module storage entry in the code cache must implement.
+pub trait ModuleStorageEntryInterface: Sized + Debug {
+    /// Given a state value, constructs a new module storage entry from it.
+    fn from_state_value(
+        runtime_environment: &RuntimeEnvironment,
+        state_value: StateValue,
+    ) -> VMResult<Self>;
+
+    /// Returns the bytes of the given module.
+    fn bytes(&self) -> &Bytes;
+
+    /// Returns the size in bytes of the given module.
+    fn size_in_bytes(&self) -> usize {
+        self.bytes().len()
+    }
+
+    /// Returns the state value metadata of the given module.
+    fn state_value_metadata(&self) -> &StateValueMetadata;
+
+    /// Returns the hash of the given module.
+    fn hash(&self) -> &[u8; 32];
+
+    /// Returns module's metadata.
+    fn metadata(&self) -> &[Metadata];
```

Comment:
I think we need to return aptos specific metadata here. i.e: aptos_framework::RuntimeModuleMetadataV1

Reply:
We still need Metadata because we use it in ResourceResolver in MoveVM? But otherwise yes, we need a way to query metadata but I on purpose left it out of this PR because it can be implemented on top of this, without bloating the PR further.